### PR TITLE
Cleanup and unify keyboard input.

### LINKED ISSUE #54797: Inconsistent output of `InputEventKey.as_text()` with Czech keyboard layout on Linux
### Godot version

3.4.stable

### System information

Linux (Manjaro)

### Issue description

It mixes "key from keyboard layout" with "physical key" (or maybe US keyboard layout?).

This is an output of me pressing top number keys from 1 to 0 (from left to right):

```
<<Plus>> 43 49
<<2>> 50 50
<<3>> 51 51
<<4>> 52 52
<<5>> 53 53
<<6>> 54 54
<<ý>> 253 55
<<á>> 225 56
<<í>> 237 57
<<é>> 233 48
```

Either everything should be a number (US keyboard layout interpretation) `1234567890` or it should be `+ěščřžýáíé` (CS keyboard layout). Now it's somehow mixed together.

### Steps to reproduce

Select a non-US keyboard layout (e.g. Czech) and call `as_text` on `InputEventKey`. Try pressing keys which are different to US layout.

### Minimal reproduction project

```gdscript
func _input(event: InputEvent) -> void:
	if event is InputEventKey && event.is_pressed():
		print("<<%s>> %s %s" % [event.as_text(), event.scancode, event.physical_scancode])
```

### LINKED ISSUE #69830: InputEventKey.get_keycode_with_modifiers() returns wrong values (German keyboard)
### Godot version

4.0 beta8

### System information

Windows 10

### Issue description

`InputEventKey.get_keycode_with_modifiers()` returns wrong values on German Keyboard.

In case it matters: This happens for my notebook's keyboard (Lenovo T470p) as well as my external Microsoft Ergonomic keyboard. The keyboards' drivers in the Device Manager are named "HID keyboard device" and "Standard PS/2 Keyboard"

### Steps to reproduce

1. Set Windows to [German keyboard layout](https://en.wikipedia.org/wiki/German_keyboard_layout) and language.
2. Make a node with the code below and run it (or use Minimal reproduction project)
3. When pressing for example the Umlaut-Keys ö, ü or ä, I get 96 (QuoteLeft), 59 (Semicolon) and 39 (Apostrophe) respectively (they are not equal to the physical keycodes though). I would expect keycodes 214 (Ö), 220 (Ü) and 196 (Ä). The locale is returned correctly however.

```
extends Control

func _ready():
	print(OS.get_locale())
	print(OS.get_locale_language())

func _input(event):
	if event is InputEventKey:
		var key: int = event.get_keycode_with_modifiers()
		var key_phys: int = event.get_physical_keycode_with_modifiers()
		print("keycode  %s (%s), physical %s (%s)" %
			[key, OS.get_keycode_string(key), 
			key_phys, OS.get_keycode_string(key_phys)])
```

### Minimal reproduction project

[get_keycode_demo.zip](https://github.com/godotengine/godot/files/10198319/get_keycode_demo.zip)

### PR DESCRIPTION
**Pull request by bruvzg**
- Unify `keycode` values (secondary label printed on a key), remove unused hard-coded Latin-1 codes.
- Unify IME behavior, add inline composition string display on Windows and X11.
- Add `key_label` (primary label printed on a key) value to the key events, and allow mapping actions to the `key_label` events.
- Add support for physical keyboard (Bluetooth, USB or Sidecar) handling on iOS.
- Remove `LineEdit`/`TextEdit` selected text when IME input begin.
- Add support for media key handling on macOS.

KeyboardTranslation

- Physical keyboard should always have `keycode`, `key_label` and `physical_keycode` set, and `unicode` set on key press only when it's relevant.
- IME only sends `unicode` key down evens, all other keys are ignored while it's active.
- Virtual keyboards send everything physical do, except `physical_keycode`.

Supported input modes:
|                   | Android   | iOS       | Linux/X11    | macOS        | Windows      | Web       |
|-------------------|-----------|-----------|--------------|--------------|--------------|-----------|
| Physical Keyboard | Yes [^2]  | Yes [^3]  | Yes          | Yes          | Yes          | Yes [^5]  |
| Virtual Keyboard  | Yes       | Yes       | No [^4]      | No [^4]      | No [^4]      | No        |
| IME               | [^1]      | [^1]      | Yes (inline) | Yes (inline) | Yes (inline) | No        |

[^1]: Using custom virtual keyboard layout.
[^2]: Tested with Bluetooth and USB-OTG keyboards.
[^3]: Tested with Bluetooth and USB-OTG keyboards, Smart Connector case keyboard and Sidecar.
[^4]: OS on-screen keyboards emulate physical keyboard.
[^5]: `key_label` values aren't supported (generated from `unicode` and `keycode` as fallback).

Fixes https://github.com/godotengine/godot/issues/69830
Fixes https://github.com/godotengine/godot/issues/54797
Supersede https://github.com/godotengine/godot/pull/53682

### CODE REVIEW

**Comment by Zireael07** [comment]
Can that picture go into official Godot docs?

**Comment by bruvzg** [comment]
> Can that picture go into official Godot docs?

In some form, I definitely plan to add it (https://github.com/godotengine/godot/pull/69751).

**Comment by bruvzg** [comment]
> Handle IME composition start / end to suppress key events (ideally I will try to make Implement on-the-spot IME support for X11 #53682 work).

Seems like position of the candidate window is determined by focus window, no client window, so creating subwindow will require it to be focused for IME to work. Candidate window position sometimes is off on the Windows as well.

So I guess the proper solution is to draw a custom candidate window as well.

**Comment by YuriSizov** [comment]
The term unshifted unicode in an input event makes it look like it has something to do with pressing Shift. Can we come up with an alternative, that is both correct and clear to laymen?

**Comment by bruvzg** [comment]
> The term unshifted unicode in an input event makes it look like it has something to do with pressing Shift

Well, it has something to do with `Shift`, as in "not affected by `Shift` (or other modifiers)", but I agree it's not the best name.

I would rename `unshifted_unicode` to `primary_label` and `keycode` to `secondary_label`, but the second part is a breaking change of a widely used API, and it probably is too late to do it for 4.0.

**Comment by YuriSizov** [comment]
Well, `keycode` is a pretty widely accepted outside of Godot too, so that would be confusing.

Maybe I'm missing what `unshifted_unicode` does in practice. It maps to `Key`, but our enum doesn't have definitions for every unicode character. So what does that new property hold? Is it actually a numeric Unicode value? And how do we expect people to use it? Maybe that'll lead us to a better name.

**Comment by bruvzg** [comment]
> It maps to Key, but our enum doesn't have definitions for every unicode character.

`Key` can hold any Unicode value, all defined special key enum values (with exception for ASCII values) are outside of Unicode range (shifted by 22 bits for this purpose, and it was always like this). And `OS.get_keycode_string()` will work with Unicode values.

**Comment by raulsntos** [comment]
Thanks a lot for implementing inline IME support for X11 :bow:, don't forget to add "Fixes https://github.com/godotengine/godot/issues/53658".

I'm getting this message in the console when deleting characters from a LineEdit:

```
Unicode parsing error, some characters were replaced with spaces: Invalid UTF-8 leading byte (83)
```

**Comment by bruvzg** [comment]
> Unicode parsing error, some characters were replaced with spaces: Invalid UTF-8 leading byte (83)
> Also the caret seems to always stay at the beginning during IME composition.

Which IME engine are you using, I suspect the issue is in `_xim_preedit_draw_callback` which assumes `chg_first` and `chg_length` are a number of full Unicode characters, but in some cases it probably might be a number of UTF-8 code points instead.

File: core/os/keyboard.cpp [hunk 1]
```diff
@@ -171,6 +161,10 @@ static const _KeyCodeText _keycodes[] = {
 	{Key::LAUNCHD               ,"LaunchD"},
 	{Key::LAUNCHE               ,"LaunchE"},
 	{Key::LAUNCHF               ,"LaunchF"},
+	{Key::GLOBE                 ,"Globe"},
+	{Key::KEYBOARD              ,"On-screen keyboard"},
+	{Key::JIS_EISU              ,"Jis Eisu"},
```
[the remark this answers is not in the document]
Comment (bruvzg) [reply core/os/keyboard.cpp]: Fixed

**Comment by raulsntos** [comment]
I'm using Fcitx5 (5.0.14).

**Comment by bruvzg** [comment]
> I'm using Fcitx5 (5.0.14).

And which input method? I have not found any difference between ibus and fcitx behavior, all input methods I have tested seems to use UTF-8 and always update the whole string, I was not able to reproduce "Invalid UTF-8 leading byte" error.

> Also the caret seems to always stay at the beginning during IME composition.

Caret should be fixed.

**Comment by raulsntos** [comment]
> And which input method? I have not found any difference between ibus and fcitx behavior, all input methods I have tested seems to use UTF-8 and always update the whole string, I was not able to reproduce "Invalid UTF-8 leading byte" error.

Testing with the latest artifact from CI, I seem to be getting it less often. It used to be on every deletion but now it's only happening every now and then. It happens even when IME is not active, both when using _English (US)_ and _Mozc_.

> Caret should be fixed.

It is :tada:

**Comment by bruvzg** [comment]
> It happens even when IME is not active, both when using English (US) and Mozc.

Hm, that's strange, if it's happening in non-IME mode it's probably something else. Just to be sure, is it reproducible with empty project, or only in the editor (for example, I usually test input with the project manager filter LineEdit, which have a lot of unrelated code attached to the text change event)?

**Comment by raulsntos** [comment]
> I usually test input with the project manager filter LineEdit

Oh, that's what I was using to test it too :smile: 

I tried other LineEdits in the editor and it happened too, but I was unable to reproduce in a running project.

**Review by reduz — APPROVED** [review]

**Comment by reduz** [comment]
Looks fantastic, great work!

File: doc/classes/@GlobalScope.xml [hunk 2]
```diff
@@ -2121,203 +2091,23 @@
...
 		</constant>
-		<constant name="KEY_DIVISION" value="247" enum="Key">
-			÷ key.
+		<constant name="KEY_JIS_EISU" value="4194418" enum="Key">
+			英数 key on Mac Keyboard.
 		</constant>
-		<constant name="KEY_YDIAERESIS" value="255" enum="Key">
-			ÿ key.
+		<constant name="KEY_JIS_KANA" value="4194419" enum="Key">
+			かな key on Mac Keyboard.
```
Comment (akien-mga) [inline doc/classes/@GlobalScope.xml:2110]: "Mac keyboard" with lower case "k", unless it's branded this way?

File: doc/classes/InputEventKey.xml [hunk 3]
```diff
@@ -5,15 +5,42 @@
...
+		<method name="as_text_keycode" qualifiers="const">
+			<return type="String" />
+			<description>
+				Returns a [String] representation of the event [member keycode] and modifiers.
+			</description>
+		</method>
+		<method name="as_text_physical_keycode" qualifiers="const">
+			<return type="String" />
+			<description>
+				Returns a [String] representation of the event [member physical_keycode] and modifiers.
```
Comment (akien-mga) [inline doc/classes/InputEventKey.xml:30]:
```suggestion
				Returns a [String] representation of the event's [member key_label] and modifiers.
			</description>
		</method>
		<method name="as_text_keycode" qualifiers="const">
			<return type="String" />
			<description>
				Returns a [String] representation of the event's [member keycode] and modifiers.
			</description>
		</method>
		<method name="as_text_physical_keycode" qualifiers="const">
			<return type="String" />
			<description>
				Returns a [String] representation of the event's [member physical_keycode] and modifiers.
```

File: doc/classes/InputEventKey.xml [hunk 4]
```diff
@@ -5,15 +5,42 @@
...
+			</description>
+		</method>
+		<method name="get_key_label_with_modifiers" qualifiers="const">
+			<return type="int" enum="Key" />
+			<description>
+				Returns the localized key label combined with modifier keys such as [kbd]Shift[/kbd] or [kbd]Alt[/kbd]. See also [InputEventWithModifiers].
+				To get a human-readable representation of the [InputEventKey] with modifiers, use [code]OS.get_keycode_string(event.get_key_label_with_modifiers())[/code] where [code]event[/code] is the [InputEventKey].
```
Comment (akien-mga) [inline doc/classes/InputEventKey.xml:37]: Not something to change in this PR, but I wonder why this helper is in `OS` and not in `Input`. Would it make sense to move it?

File: doc/classes/InputEventKey.xml [hunk 5]
```diff
@@ -29,19 +56,36 @@
 		<member name="echo" type="bool" setter="set_echo" getter="is_echo" default="false">
 			If [code]true[/code], the key was already pressed before this event. It means the user is holding the key down.
 		</member>
+		<member name="key_label" type="int" setter="set_key_label" getter="get_key_label" enum="Key" default="0">
+			Represent localized label printed on the key in the current keyboard layout, which corresponds to one of the [enum Key] constants or any valid Unicode character.
+			For keyboard layout with the single label equal to [member keycode].
```
Comment (akien-mga) [inline doc/classes/InputEventKey.xml:61]:
```suggestion
			Represents the localized label printed on the key in the current keyboard layout, which corresponds to one of the [enum Key] constants or any valid Unicode character.
			For keyboard layouts with a single label on the key, it is equivalent to [member keycode].
```

File: doc/classes/InputEventKey.xml [hunk 6]
```diff
@@ -29,19 +56,36 @@
...
+			[codeblock]
+			    +-----+ +-----+
+			    | Q   | | Q   | - "Q" - keycode
+			    |   Й | |  ض | - "Й" and "ض" - key_label
+			    +-----+ +-----+
+			[/codeblock]
 		</member>
 		<member name="physical_keycode" type="int" setter="set_physical_keycode" getter="get_physical_keycode" enum="Key" default="0">
-			Key physical keycode, which corresponds to one of the [enum Key] constants. Represent the physical location of a key on the 101/102-key US QWERTY keyboard.
+			Represent the physical location of a key on the 101/102-key US QWERTY keyboard, which corresponds to one of the [enum Key] constants.
```
Comment (akien-mga) [inline doc/classes/InputEventKey.xml:81]:
```suggestion
			Represents the physical location of a key on the 101/102-key US QWERTY keyboard, which corresponds to one of the [enum Key] constants.
```

File: doc/classes/InputEventKey.xml [hunk 7]
```diff
@@ -29,19 +56,36 @@
...
 			To get a human-readable representation of the [InputEventKey], use [code]OS.get_keycode_string(event.keycode)[/code] where [code]event[/code] is the [InputEventKey].
 		</member>
 		<member name="pressed" type="bool" setter="set_pressed" getter="is_pressed" default="false">
 			If [code]true[/code], the key's state is pressed. If [code]false[/code], the key's state is released.
 		</member>
 		<member name="unicode" type="int" setter="set_unicode" getter="get_unicode" default="0">
-			The key Unicode identifier (when relevant). Unicode identifiers for the composite characters and complex scripts may not be available unless IME input mode is active. See [method Window.set_ime_active] for more information.
+			The key Unicode character code (when relevant), shifted by modifier keys. Unicode character code for the composite characters and complex scripts may not be available unless IME input mode is active. See [method Window.set_ime_active] for more information.
```
Comment (akien-mga) [inline doc/classes/InputEventKey.xml:88]:
```suggestion
			The key Unicode character code (when relevant), shifted by modifier keys. Unicode character codes for composite characters and complex scripts may not be available unless IME input mode is active. See [method Window.set_ime_active] for more information.
```

File: editor/action_map_editor.cpp [hunk 8]
```diff
@@ -453,10 +453,14 @@ void ActionMapEditor::update_action_list(const Vector<ActionInfo> &p_action_info
 			// First Column - Icon
 			Ref<InputEventKey> k = event;
 			if (k.is_valid()) {
-				if (k->get_physical_keycode() == Key::NONE) {
+				if (k->get_physical_keycode() == Key::NONE && k->get_keycode() == Key::NONE && k->get_key_label() != Key::NONE) {
+					event_item->set_icon(0, action_tree->get_theme_icon(SNAME("KeyboardUnicode"), SNAME("EditorIcons")));
```
Comment (akien-mga) [inline editor/action_map_editor.cpp:457]: Should the icon be renamed to `KeyboardKeyLabel` for consistency?

**Review by akien-mga — APPROVED** [review]
Looks great! Will likely need extensive testing, especially for things like Android with virtual and physical keyboards, and the Web, but this will happen in the next beta :)

File: editor/action_map_editor.cpp (hunk 8 above)
Comment (bruvzg) [reply editor/action_map_editor.cpp]: Renamed.

**Comment by bruvzg** [comment]
> Unicode parsing error, some characters were replaced with spaces: Invalid UTF-8 leading byte (83)

I have checked it again, seems like `XLookupString` can return any garbage not UTF-8, so I have no idea how to get `key_label` on X11 properly. Removed it from X11 DisplayServer for now.

**Comment by bruvzg** [comment]
> Removed it from X11 DisplayServer for now.

Reimplemented it in another way, additionally use `xkbcommon` library (optionally), but should work with any system encoding.

**Comment by akien-mga** [comment]
Linux CI likely needs this change:
```diff
diff --git a/.github/workflows/linux_builds.yml b/.github/workflows/linux_builds.yml
index 84fdb2180b..ce4f757a19 100644
--- a/.github/workflows/linux_builds.yml
+++ b/.github/workflows/linux_builds.yml
@@ -85,7 +85,7 @@ jobs:
           sudo apt-get install build-essential pkg-config libx11-dev libxcursor-dev \
               libxinerama-dev libgl1-mesa-dev libglu-dev libasound2-dev libpulse-dev \
               libdbus-1-dev libudev-dev libxi-dev libxrandr-dev yasm xvfb wget unzip \
-              llvm libspeechd-dev speech-dispatcher fontconfig libfontconfig-dev
+              llvm libspeechd-dev speech-dispatcher fontconfig libfontconfig-dev libxkbcommon-dev
 
       - name: Setup Godot build cache
         uses: ./.github/actions/godot-cache
```

**Comment by bruvzg** [comment]
> This might also be needed for NixOS which puts its includes in non-standard directories:

Added. I guess I should do https://github.com/godotengine/godot/pull/71263 for the X11 libs to avoid this kind of issues.

**Comment by akien-mga** [comment]
Tested on Linux with en_US, fr_FR and da_DK keyboard layouts, seems to work great! Some tests with `event.as_text_key_label()`:

<details>
<summary>French AZERTY, pressing all keys from top to bottom (remapping, on a physical US QWERTY):</summary>

```
²
Ampersand
É
QuoteDbl
Apostrophe
ParenLeft
Minus
È
UnderScore
Ç
À
ParenRight
Equal
A
Z
E
R
T
Y
U
I
O
P
BraceLeft
Dollar
Asterisk
Q
S
D
F
G
H
J
K
L
M
Ù
W
X
C
V
B
N
Comma
Semicolon
Colon
Exclam
```

</details>

<details>
<summary>The same on Danish keyboard layout:</summary>

```
½
1
2
3
4
5
6
7
8
9
0
Plus
Equal
Q
W
E
R
T
Y
U
I
O
P
Å
BraceRight
Apostrophe
A
S
D
F
G
H
J
K
L
Æ
Ø
Z
X
C
V
B
N
M
Comma
Period
Minus
```

</details>

I noticed a difference in the way modifiers are handled, but this may be intentional. Testing on French AZERTY, where:
- Physical key "2" has key label "É", and "2" is the "shifted" key.
- Physical key "E" has key label "E", and "€" is accessible via Alt Gr + E (on Linux with the default `fr` mapping).

Pressing <kbd>Shift+É</kbd> to get "2" prints:
```
Shift
Shift+É
```

Pressing <kbd>Alt Gr+E</kbd> to get "€" prints:
```
Alt
€
```

**Comment by akien-mga** [comment]
Thanks!

**Comment by vmedea** [comment]
I'm sorry but I think the change in of `SUPER_*` and `HYPER_*` constants is a bit weird here.

From a X11/Wayland perspective the old situation makes more sense? Like, at least the left and right SUPER ("windows") key are really separate keys. I'm not sure about HYPER.

Also there's a stale `SUPER_R`/`SUPER_L` in `core/keyboard.h`, if you really intend to remove them.

**Comment by bruvzg** [comment]
> I'm sorry but I think the change in of SUPER_* and HYPER_* constants is a bit weird here.

None of the other platforms distinguish L/R Super, all use `META`, and no other key have Left/Right variant. It is possible to get L/R variants on other platforms, but if it's done it should be done for all keys (Ctrl, Alt, etc.) not just for Super/Meta and essentially nonexistent Hyper. There was PR to do it, but it was rejected.

> Also there's a stale SUPER_R/SUPER_L in core/keyboard.h, if you really intend to remove them.

Indeed, missed them, should be removed.

**Comment by vmedea** [comment]
Ok, thanks, it wasn't clear to me that META was the same thing. I thought the key was entirely removed. Makes more sense then.

**Comment by Vivalzar** [comment]
> I noticed a difference in the way modifiers are handled, but this may be intentional. Testing on French AZERTY, where:
> - Physical key "2" has key label "É", and "2" is the "shifted" key.

Yes, this is intentional. All numbers on French keyboards are accessed using `shift`.

**Comment by h0lley** [comment]
`InputEventKey` does not necessarily have the `key_label` set, such as when it has been retrieved from `InputMap.action_get_events()`. In that case, how do we obtain the localized key label of an unicode char that has no keycode representation (if it had, we could use `DisplayServer.keyboard_get_keycode_from_physical()`). Example: umlauts.

apologies if that is not within the scope of this PR. didn't find any other place related to `key_label`.

**Pull request merged**